### ISSUE TITLE
Building roads could be easier

### ISSUE THREAD

**Issue by Poppaluu** [widelands/widelands#5617]
The problem:
The biggest problem I have with this game is that the building's roads start always at the same spot. I had this building right next to a road and the flag was like the next block but since the road starts from the other side I couldn't build the road to that flag.

What I would like to happen:
I would like that you could start building the road to whichever way you want from the building. It would make city planning a lot easier and my cities would look a lot nicer

Solution:
I don't know much about coding with c++ or coding games but I would think it would be quite easy to just make the buildings be flags also and remove the first road built automatically by the structure. I don't even think that you need to rotate the building or make new assets for rotated buildings, since the help this would bring would be much more than the possible eyesore of houses having roads leave from behind them.

Another nice addition to building roads would be that there would be a "ghost" road when building the road showing you where the road would be built

**Comment by klaus-halfmann** [comment]
Hello @Poppaluu  I am not sure what your Problem is, I think we have a _lot_ of ways to build Roads.
Please check my Videos at https://www.youtube.com/channel/UCHHf08zkQGFasRqgyznCC-A or some other Videos.
As I am stuck at home with Corona anyway I would even offer son Online Sessions with a video call so I can show
you the Tricks you did not yet find in the Tutorials.

If you still think this needs Improvement, maybe record a video showing what you want to achieve?

**Comment by RHolton** [comment]
I think Poppaluu doesn't like that each building starts with what appears to be a road segment, and all those initial "road segments" point in the same direction. He would like to be able to connect roads to buildings from any direction -- not having to connect to the end of that initial "road segment".

I'd say it would be cool if you could rotate buildings before you place them so that the initial "road segment" points in any direction. 

I realize that such a change would be quite a bit of work -- just thinking of all the new animations required gives me a headache. 

But then it would also "be cool" if you could rotate the map.

But I'm not holding my breath for any of that.

**Comment by Noordfrees** [comment]
Rotations were previously discussed and shot down in https://www.widelands.org/forum/topic/4902/?page=2 . This kind of effort is not practically doable.

Also, any such change will have a **huge** impact on the game mechanics, may unbalance maps, and would break a lot of assumptions in the code about the layout of buildings and flags. So this is neither easy to do nor a change that can be performed without broad discussion and consensus from the playerbase. Planning your settlements around the flag-placement restriction is a very important strategical aspect of Widelands.

**Comment by Poppaluu** [comment]
I was just thinking that the buildings would be placed without the initial road exactly as @RHolton said. I thought the implementation wouldn't need any re-animations if you just don't have the initial road be built and make the buldings act as flags since you can connect roads only to the flags. this would make it possible to start the road building to any direction. I don't think it would matter if the building would be facing a different direction than the road is exiting from the building since this would be a much-needed fix in my opinion. @klaus-halfmann

**Comment by Poppaluu** [comment]
![widelands_road_building_problem](https://user-images.githubusercontent.com/98604428/196941985-7e6a268b-9223-4c6e-8483-16349ab8faaf.png)

I find it unreasonable that I cannot connect these two flags @Noordfrees. Instead of it being a strategy part of the game I feel like it comes off as unnatural and taking away from the experience of the game

**Comment by Noordfrees** [comment]
If you made buildings act like flags, that would mean that all roads would lead right through buildings, entering through one wall and leaving through the other wall. Or how do you mean?

I don't see anything wrong with this image. Delete the existing road (the one the mouse is over), build a road between the two flags, rebuild the old road from the new (lower) flag to the destination east of the image.

**Comment by Poppaluu** [comment]
@Noordfrees You're right that the road going through the house would be quite silly. What if buildings could have only one road built from them then?
I just find it more pretty on the eyes if I wouldn't have to have a piece of road I don't need in my city. In the picture it could just go straight to the flag next to it but now I need extra steps to make it work

**Comment by Noordfrees** [comment]
Even then it would require every building to have six different sets of images for all possible rotations. Imagine the sentry above had a road leading to it from the west. The soldiers entering and leaving the building would be walking through the wall. And anything involving rotated graphics is just not possible due to the amount of graphics design work involved.

**Comment by tothxa** [comment]
> I find it unreasonable that I cannot connect these two flags @Noordfrees. Instead of it being a strategy part of the game I feel like it comes off as unnatural and taking away from the experience of the game

I disagree, and this actually strengthened my opinion against building and map rotation. Allowing multiple buildings connect to a single flag would allow much tighter packed economies, so that would be a huge change for the strategy of the game.

I realise that the current design comes from a >25 year old optimisation for a game written for computers with nearly 3 orders of magnitude lower specs in all regards than today's and completely lacking 3D hardware. But then you might as well question why you even need flags. And then soon we find ourselves discussing a game of an entirely different class.

**Comment by tothxa** [comment]
> Even then it would require every building to have six different sets of images for all possible rotations.

Even bigger question is what to do with big buildings? Rotating them would change their space requirement too.

**Comment by frankystone** [comment]
> ![widelands_road_building_problem](https://user-images.githubusercontent.com/98604428/196941985-7e6a268b-9223-4c6e-8483-16349ab8faaf.png)
> 
> I find it unreasonable that I cannot connect these two flags @Noordfrees. Instead of it being a strategy part of the game I feel like it comes off as unnatural and taking away from the experience of the game

Long time ago i read about the possibility in the Settlers series to "move" a road onto a flag. Say the Flag action menu of the  lumberacks hut has an additional entry _"connect with nearby road"_ and the algorithm is doing what Nordfriese said:

> Delete the existing road (the one the mouse is over), build a road between the two flags, rebuild the old road from the new (lower) flag to the destination east of the image.  

Don't know how much effort this will be though.

See the last paragraph of this post: https://www.widelands.org/forum/topic/1074/

**Comment by Noordfrees** [comment]
> Long time ago i read about the possibility in the Settlers series to "move" a road onto a flag. Say the Flag action menu of the lumberacks hut has an additional entry "connect with nearby road" and the algorithm is doing what Nordfriese said:

That sounds like a very useful feature! Should not be hard for the general case, though the corner cases may get … interesting. It also only can be done reliably when at least two nodes of the same road are adjacent to the flag we want to merge the road with.

In such situations I now often build a temporary road around, then refactor the roads, then demolish the temp road so as to avoid an economy split and merge. Doing this automatically would be great.

How would it behave when there are multiple roads nearby? Should we merge *all* candidate roads with the flag, or have separate buttons (up to 3) "merge with north-eastern road", "merge with south-western road", etc?

**Comment by frankystone** [comment]
Maybe this can be applied to the road action window? Ideally i think a workflow like this will be useful:

1. Click on road (openes road action window) The opened window will have a button "connect with nearby flag"
2. Clicking on the button Connect with flag let the window disappear and the cursor changes
3. Click on Flag

Now the road get automatically connected with the flag one has clicked on.

This way the player can choose if two roads are nearby and the algorithm doesn't have to "think" about it.

**Comment by Noordfrees** [comment]
-1 for anything involving a new mouse mode. Anything that involves non-default actions when clicking on the map other than default mode and roadbuilding mode is essentially a permanent workaround in bad coding style. Adding more of them is fragile and ugly. If we want to do that we'd need to refactor the IBase input handling system to truly support arbitrarily many different mouse modes. So purely from the implementation side please let's not do that without pressing need. Also this workflow has one extra click in it compared to my suggestion.

---

Re the forum thread, +1 for being able to remove a flag that has exactly two roads attached without deleting the roads. Should be an extra button in the flag's fieldaction window.

**Comment by tothxa** [comment]
> Re the forum thread, +1 for being able to remove a flag that has exactly two roads attached without deleting the roads. Should be an extra button in the flag's fieldaction window.

Actually (at least) in the original Settlers, the destroy flag option only destroyed the flag, and left the road. Destroying the road was a separate step. I think that makes more sense than forcing destroy road on destroy flag. Though in Settlers the icon changed under the mouse, so it was just a second click on the same spot if you wanted to do both.
